interoperability with the Jakarta Concurrency specification (#1003)
jakartaee/data: spec/src/main/asciidoc/repository.asciidoc

- `R m(P1 p1, P2 p2, ...)` is a legal repository method signature according to the previous sections of this chapter, or, in the special case that `R` is the type `java.lang.Void`, `void m(P1 p1, P2 p2, ...)` is a legal repository method signature according to the previous sections of this chapter, and
- `F` is a parameterized type representing a value which might not yet have been computed and which is supported by the Jakarta Data provider.

> NOTE: Repository methods annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification must be able to return the type `java.util.concurrent.CompletionStage`, in which case the Jakarta Data provider returns an already completed `CompletionStage` such that the Jakarta Concurrency provider is able to control the asynchronous behavior. Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.

Reviewer: ```suggestion
NOTE: Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.

NOTE: A repository method annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification is permitted to declare the return type `java.util.concurrent.CompletionStage`. 
In this case, the Jakarta Data provider must return an already-completed `CompletionStage` so that the Jakarta Concurrency provider is able to control the asynchronous behavior.
```

Reviewer: There's a bit of a hole here; `@Asynchronous` methods don't (I assume) propagate the transaction context. So we should say something about what transaction context they run in. Does the async interceptor get applied before or after the tx management interceptor?

Reviewer: Ah, there, no hole, found this:

> The Jakarta EE Product Provider must assign the interceptor for asynchronous methods to have priority of `Interceptor.Priority.PLATFORM_BEFORE + 5`. Interceptors with a lower priority, such as `Transactional`, must run on the thread where the asynchronous method executes, rather than on the submitting thread. When an asynchronous method is annotated as `Transactional`, the transactional types which can be used are: `TxType.REQUIRES_NEW`, which causes the method to run in a new transaction, and `TxType.NOT_SUPPORTED`, which causes the method to run with no transaction. All other transaction attributes must result in [UnsupportedOperationException](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/UnsupportedOperationException.html?is-external=true) upon invocation of the asynchronous method.

That's fine.

> NOTE: Repository methods that are annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification must be able to return the types `java.util.concurrent.CompletionStage` and `java.util.concurrent.CompletableFuture`, in which case the Jakarta Data provider returns an already completed `CompletionStage` or `CompletableFuture` such that the Jakarta Concurrency provider is able to control the asynchronous behavior. Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.

Reviewer: ```suggestion
NOTE: Repository methods annotated with the `@Asynchronous` annotation from the Jakarta Concurrency specification must be able to return the types `java.util.concurrent.CompletionStage` and `java.util.concurrent.CompletableFuture`, in which case the Jakarta Data provider returns an already completed `CompletionStage` or `CompletableFuture` such that the Jakarta Concurrency provider is able to control the asynchronous behavior. Every Jakarta Data provider is otherwise encouraged, but not required, to support asynchronous repository methods returning `java.util.concurrent.CompletionStage`.
```

Reviewer: I'm not sure we should allow `CompletableFuture` here. I think `CompletionStage` is enough.

Author: The Jakarta Concurrency spec explicitly states `CompletableFuture` as one of the 3 return types to use with [@Asynchronous](https://jakarta.ee/specifications/concurrency/3.1/apidocs/jakarta.concurrency/jakarta/enterprise/concurrent/asynchronous), so it seems correct to mention it:

> Methods with the following return types can be annotated to be asynchronous methods:
>
>  * [CompletableFuture](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletableFuture.html)
>  * [CompletionStage](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/CompletionStage.html)
>  * void

Reviewer: > The Jakarta Concurrency spec explicitly states CompletableFuture as one of the 3 return types

Yes, sure, I saw that, but that doesn't mean that Jakarta Data should require that providers support this.

And, I'm not an expert on this stuff, but it seems to me that it's better to _not_ expose the `complete()` method to repository clients. It's the responsibility of the repository implementation to call this operation, and clients should not be messing with it. 

AFAIK, `CompletionStage` is the right level of abstraction for clients.

Author: > it seems to me that it's better to _not_ expose the `complete()` method to repository clients

Yeah, I agree. I was only thinking about it relation to Find operations. But if a user has an asynchronous update/delete/insert, we wouldn't want them to be able to forcibly complete or cancel its CompletableFuture because it would be uncertain to the application whether the change was made or not.  I added commit https://github.com/jakartaee/data/pull/1003/commits/79c9fa5be19bb4717ed28471c949e9436fe7049c to remove CompletableFuture.

An _asynchronous repository_ is a repository which declares asynchronous repository methods.

> A mixture of synchronous and asynchronous repository methods can be used within the same repository if all of the asynchronous methods are annotated with the `@Asynchronous` annotation to have the Jakarta Concurrency provider handle the asynchronous behavior. Otherwise, the Jakarta Data provider is not required to support mixing synchronous and asynchronous repository methods within the same repository interface.

Reviewer: I would like to say something here which allows _reactive_ repositories to _not_ have to support `@Asynchronous`, because it would be some sort of weird noop for them.

Author: Yes, that makes sense.  How about this wording?  https://github.com/jakartaee/data/pull/1003/commits/818dfd8c98414fc7029a9e01b609e1889ce8fa90  It disallows `@Asynchronous` in that case.

> A mixture of synchronous and asynchronous repository methods can be used within the same repository if all of the asynchronous methods are annotated with the `@Asynchronous` annotation to have the Jakarta Concurrency provider handle the asynchronous behavior. Otherwise, the Jakarta Data provider is not required to support mixing synchronous and asynchronous repository methods within the same repository interface. The `@Asynchronous` annotation must not be used on reactive repositories.

Reviewer: ```suggestion
A mixture of synchronous and asynchronous repository methods can be used within the same repository if all of the asynchronous methods are annotated with the `@Asynchronous` annotation to have the Jakarta Concurrency provider handle the asynchronous behavior. Otherwise, the Jakarta Data provider is not required to support mixing synchronous and asynchronous repository methods within the same repository interface. The `@Asynchronous` annotation must not be used on repositories implemented using reactive streams.
```

> A mixture of synchronous and asynchronous repository methods can be used within the same repository if all of the asynchronous methods are annotated with the `@Asynchronous` annotation to have the Jakarta Concurrency provider handle the asynchronous behavior. Otherwise, the Jakarta Data provider is not required to support mixing synchronous and asynchronous repository methods within the same repository interface. The `@Asynchronous` annotation must not be used on repositories implemented using reactive streams.

Reviewer: ```suggestion
A repository may declare a mixture of synchronous and asynchronous repository methods if every asynchronous method is annotated with the `@Asynchronous` annotation, so that Jakarta Concurrency provides the asynchronous behavior. 
Otherwise, the Jakarta Data provider is not required to support mixing synchronous and asynchronous repository methods within the same repository interface. 
The `@Asynchronous` annotation must not be used on repositories implemented using reactive streams.
```
